Fix parallel states with long ID dec or name

terminalmage opened the pull request:
The cache files used for these are based on the state's tag, so longer ID decs or names will cause errors when the filename's length exceeds the max allowed by the kernel.

This fixes these errors by taking (up to) the first 32 chars of the result of base64-encoding the tag, and using that as the parallel cache filename.

Fixes #49738.

terminalmage committed:
Fix parallel states with long ID dec or name

The cache files used for these are based on the state's tag, so longer
ID decs or names will cause errors when the filename's length exceeds
the max allowed by the kernel.

This fixes these errors by taking (up to) the first 32 chars of the
result of base64-encoding the tag, and using that as the parallel cache
filename.

terminalmage commented:
I found some other weirdness with the test suite. The `setUp` should have been a `setUpClass`, and there is a `tearDown` that only impacts a single test and should not be run after every test, but as the `finally` portion of a `try/finally` block.

terminalmage committed:
Remove/replace unnecessary setUp/tearDown

The setUp only needs to run at the beginning so it can be a setUpClass,
and the tearDown logic only applied to a single test and should not be
run after each test.

terminalmage commented:
OK, I've push another commit that replaces the setUp/tearDown.

rallytime commented:
@terminalmage It appears the windows tests didn't like this change: https://jenkinsci.saltstack.com/job/pr-windows2016-py2/job/PR-49777/2/

Can you take a look?

terminalmage committed:
Add additional info to the assert to aid in troubleshooting

terminalmage committed:
Skip parallel states test on Windows until we can get it to work

terminalmage commented:
@rallytime the failing test on Windows was due to parallel states apparently not working on Windows. I have skipped the failing test on Windows only for the meantime, and I've opened https://github.com/saltstack/salt/issues/49808 to track getting parallel states working.

terminalmage committed:
Explicitly import salt.utils.hashutils

The code works without this because at least one of the other modules
being imported also imports this. But explicitly importing it here will
keep us from getting bitten if the imports change elsewhere later on.

terminalmage commented:
I added one last commit to make the `salt.utils.hashutils` import explicit.

damon-atkins commented:
Won't two commands with the same first 22 characters (32 encoded) use the same file, within the same job id?

```
test_cmd_tag_check_dog:
  cmd.run:
    - name: 'echo quick brown fox jumped over the lazy dog'
    - parallel: True
 test_cmd_tag_check_sheep:
  cmd.run:
    - name: 'echo quick brown fox jumped over the lazy sheep'
    - parallel: True
```

How about "SLS ID" and documenting they have a length limit determined by the file system.

terminalmage committed:
Switch from base64 encoding to sha1 digest for unique cache filename

terminalmage commented:
Absolutely not, that is just an arbitrary limit.

I've switched from base64 to SHA1. A SHA1 hash of the tag should provide a short enough (while still remaining reasonably unique) filename.

damon-atkins commented:
Is it worth added the SLS ID into the string as well before generating SHA1 to cover someone running the same command string parallel or something like this?

damon-atkins commented on tests/integration/modules/test_state.py:
```diff
@@ -1423,20 +1406,48 @@ def test_state_nonbase_environment(self):
+        parallel state cache were previously based on the tag for each chunk,
+        and longer ID decs or name params can cause the cache file to be longer
+        than the operating system's max file name length. To counter this we
+        instead base64-encode the chunk's tag (as this is a bit faster than
```
Suggest this needs some rewording at some stage.

terminalmage commented:
>Is it worth added the SLS ID into the string as well before generating SHA1 to cover someone running the same command string parallel or something like this?

No. The tag which is being hashed is already unique because it includes the ID declaration and `name` param, and the ID dec is guaranteed to be unique (or else the SLS file fails to render due to conflicting ID).

terminalmage commented:
And actually, that means that the base64-encoded result should be unique as well for that matter. But SHA1 should still be less likely to collide.

terminalmage committed:
Update test docstring to reflect switch to using SHA1

damon-atkins commented on salt/utils/hashutils.py:
```diff
@@ -82,6 +82,16 @@ def md5_digest(instr):
 
+def sha1_digest(instr):
+    '''
+    Generate an sha1 hash of a given string.
```
Might want to add a comment "sha1 is not consider secure, however an acceptable use is for generating fix sized filenames". Maybe md5 should have the same comment as well.   Just a thought.
